thesession.org: [content-visibility] SVG text not shown in content-visibility: auto subtree.

sammygill opened the pull request:
#### 84a4a18b9d5d794f8cc440c2640ab2dc4851789e
<pre>
thesession.org: [content-visibility] SVG text not shown in content-visibility: auto subtree.
<a href="https://bugs.webkit.org/show_bug.cgi?id=281570">https://bugs.webkit.org/show_bug.cgi?id=281570</a>
<a href="https://rdar.apple.com/138040315">rdar://138040315</a>

Reviewed by Alan Baradlay.

RenderSVGText expects to be able to perform some sort of initialization on its first pass
of layout by checking everHadLayout(). Unfortunately, if this renderer is in a
content-visibility: auto subtree, the content-visibility logic will descend down that
subtree, clearing layout on all the renderers, which ends up setting the &quot;ever had layout,&quot;
bit. This ends up confusing RenderSVGText since it thinks it has gone through layout and
assumes that it had performed this initialization, which can result in us not computing
the proper geometry for the content.

To address this, let&apos;s keep track of a bit on RenderSVGText that indicates if it has ever
performed a full pass of layout. We can use this in the same places we were previously
using everHadLayout().

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-svg-text-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-svg-text.html: Added.
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::shouldHandleSubtreeMutations const):
(WebCore::RenderSVGText::subtreeTextDidChange):
(WebCore::RenderSVGText::layout):
* Source/WebCore/rendering/svg/RenderSVGText.h:

Canonical link: <a href="https://commits.webkit.org/286447@main">https://commits.webkit.org/286447@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4354e1319f0be0da5aec2f448a417f3922c6b5d0

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/76001 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/55030 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/28900 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/80498 "Built successfully") | [✅ 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/27267 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/78117 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/64172 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/3326 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/59593 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/17747 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/79068 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/49477 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/65268 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/39951 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/46876 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/64/builds/22752 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/25594 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/67992 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/23089 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/81962 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/3370 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/2151 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/67821 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/3524 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/65236 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/67131 "Found 1 new API test failure: /TestWebKit:WebKit.WKPageGetScaleFactorNotZero (failure)") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/16726 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/11080 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/9200 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/3320 "Built successfully") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/6126 "Built successfully") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/3341 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/4279 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/124/builds/3348 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

alanbaradlay commented on Source/WebCore/rendering/RenderObject.cpp:
```diff
@@ -569,8 +569,9 @@ static inline bool objectIsRelayoutBoundary(const RenderElement* object)
-    setEverHadLayout();
+    // FIXME: Consider expending the logic to not set the "ever had layout" bit to true when "hadSkippedLayout"
+    if (!isRenderSVGText() || hadSkippedLayout == HadSkippedLayout::No)
+        setEverHadLayout();
```
This logic should either be applied to all renders or none of them. While RenderSVGText may have some internal logic that relies on this specific bit, such mismatching state may confuse the rest of the rendering code.
If RenderSVGText needs to know if we ever called layout on it, it should track it internally instead.

alanbaradlay commented on Source/WebCore/rendering/svg/RenderSVGText.h:
```diff
@@ -105,6 +105,7 @@ class RenderSVGText final : public RenderSVGBlock {
     bool m_needsPositioningValuesUpdate : 1 { false };
     bool m_needsTransformUpdate : 1 { true }; // FIXME: [LBSE] Only needed for legacy SVG engine.
     bool m_needsTextMetricsUpdate : 1 { false };
+    bool m_performedFullLayoutOnce : 1 { false }; // Needed to distinguish between when we perform a full pass of layout and everHadLayout (which can be set be content visibility for skipped content).
```
m_hasPerformedLayout ?

alanbaradlay replied:
probably whatever this renderer is doing when m_hasPerformedLayout is false should be moved to a more suitable place (e.g. insertedIntoTree etc) 🤷‍♂️

sammygill committed:
thesession.org: [content-visibility] SVG text not shown in content-visibility: auto subtree.
https://bugs.webkit.org/show_bug.cgi?id=281570
rdar://138040315

Reviewed by Alan Baradlay.

RenderSVGText expects to be able to perform some sort of initialization on its first pass
of layout by checking everHadLayout(). Unfortunately, if this renderer is in a
content-visibility: auto subtree, the content-visibility logic will descend down that
subtree, clearing layout on all the renderers, which ends up setting the "ever had layout,"
bit. This ends up confusing RenderSVGText since it thinks it has gone through layout and
assumes that it had performed this initialization, which can result in us not computing
the proper geometry for the content.

To address this, let's keep track of a bit on RenderSVGText that indicates if it has ever
performed a full pass of layout. We can use this in the same places we were previously
using everHadLayout().

* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-svg-text-expected.html: Added.
* LayoutTests/imported/w3c/web-platform-tests/css/css-contain/content-visibility/content-visibility-auto-svg-text.html: Added.
* Source/WebCore/rendering/svg/RenderSVGText.cpp:
(WebCore::RenderSVGText::shouldHandleSubtreeMutations const):
(WebCore::RenderSVGText::subtreeTextDidChange):
(WebCore::RenderSVGText::layout):
* Source/WebCore/rendering/svg/RenderSVGText.h:

Canonical link: https://commits.webkit.org/286447@main
